runtime/service/src/main/java/org/apache/polaris/service/storage/s3/sign/S3RemoteSigningCatalogHandler.java

```diff
@@ -0,0 +1,198 @@
+    throwIfRemoteSigningNotEnabled(callContext.getRealmConfig(), catalogEntity);
+
+    var result =
+        InMemoryStorageIntegration.validateAllowedLocations(
+            callContext.getRealmConfig(),
+            getAllowedLocations(tableIdentifier),
+            getStorageActions(s3SignRequest),
+            getTargetLocations(s3SignRequest));
+
+    if (result.values().stream().anyMatch(r -> r.values().stream().anyMatch(v -> !v.isSuccess()))) {
+      throw new ForbiddenException("Requested S3 location is not allowed.");
+    }
+
+    PolarisS3SignResponse s3SignResponse = s3RequestSigner.signRequest(s3SignRequest);
+    LOGGER.debug("S3 signing response: {}", s3SignResponse);
+
+    return s3SignResponse;
+  }
+
+  private Collection<String> getAllowedLocations(TableIdentifier tableIdentifier) {
+
+    if (baseCatalog.tableExists(tableIdentifier)) {
+
+      // If the table exists, get allowed locations from the table metadata
+      Table table = baseCatalog.loadTable(tableIdentifier);
+      if (table instanceof BaseTable baseTable) {
+        return StorageUtil.getLocationsUsedByTable(baseTable.operations().current());
+      }
+
+      throw new ForbiddenException("No storage configuration found for table.");
+
+    } else {
+
+      // If the table or view doesn't exist, the engine might be writing the manifests before the
+      // table creation is committed. In this case, we still need to check allowed locations from
+      // the parent entities.
+
+      PolarisResolvedPathWrapper resolvedPath =
+          CatalogUtils.findResolvedStorageEntity(resolutionManifest, tableIdentifier);
+
+      Optional<PolarisEntity> storageInfo = FileIOUtil.findStorageInfoFromHierarchy(resolvedPath);
+
+      var configurationInfo =
+          storageInfo
+              .map(PolarisEntity::getInternalPropertiesAsMap)
+              .map(info -> info.get(PolarisEntityConstants.getStorageConfigInfoPropertyName()))
+              .map(PolarisStorageConfigurationInfo::deserialize);
+
+      if (configurationInfo.isEmpty()) {
+        throw new ForbiddenException("No storage configuration found for table.");
+      }
+
+      return configurationInfo.get().getAllowedLocations();
+    }
+  }
+
+  private Set<PolarisStorageActions> getStorageActions(PolarisS3SignRequest s3SignRequest) {
+    // TODO M2: better handling of DELETE and LIST
+    return s3SignRequest.write()
+        ? Set.of(PolarisStorageActions.WRITE)
+        : Set.of(PolarisStorageActions.READ);
+  }
+
+  private Set<String> getTargetLocations(PolarisS3SignRequest s3SignRequest) {
+    // TODO M2: map http URI to s3 URI
+    return Set.of();
+  }
+
+  public static void throwIfRemoteSigningNotEnabled(
+      RealmConfig realmConfig, CatalogEntity catalogEntity) {
+    if (catalogEntity.isExternal()) {
+      throw new ForbiddenException("Remote signing is not enabled for external catalogs.");
+    }
```

Comment:
I understand we discussed this before for federated catalog we should disable remote signing but during that time we didn;t support cred vending for them either, i wonder if its  a good time to reconsider that remote signing would be possible too (ofc in a new PR)

Reply:
Totally agree, I think we could discuss this possibility later on.